File: 2023/03/ts/main.ts
import { run } from 'src/run.ts';

/** If part 2 test input has completely different input, set this to `true`. */
export const HAS_ALTERNATE = false;

function isNum(c: string) {
   return (
      c === '0' ||
      c === '1' ||
      c === '2' ||
      c === '3' ||
      c === '4' ||
      c === '5' ||
      c === '6' ||
      c === '7' ||
      c === '8' ||
      c === '9'
   );
}

function isSymbol(c: string) {
   return (
      c === '*' ||
      c === '$' ||
      c === '=' ||
      c === '#' ||
      c === '%' ||
      c === '/' ||
      c === '&' ||
      c === '+' ||
      c === '-' ||
      c === '@'
   );
}

function yeetTheNumber(grid: string[][], x: number, y: number): string {
   let res = '';
   if (isNum(grid[y][x])) {
      res += grid[y][x];
      grid[y][x] = '.';

      if (x > 0) res = yeetTheNumber(grid, x - 1, y) + res;
      if (x < grid[y].length - 1) res += yeetTheNumber(grid, x + 1, y);
   }
   return res;
}

export function part1(input: string, _isTest: boolean): string {
   const grid = input.split('\n').map((s) => s.split(''));
   const SZ = grid.length;

   let res = 0;

   for (let y = 0; y < SZ; y++) {
      for (let x = 0; x < SZ; x++) {
         if (isSymbol(grid[y][x])) {
            if (x > 0) {
               if (y < SZ - 1) res += Number(yeetTheNumber(grid, x - 1, y + 1));
               if (y > 0) res += Number(yeetTheNumber(grid, x - 1, y - 1));
               res += Number(yeetTheNumber(grid, x - 1, y));
            }
            if (x < SZ - 1) {
               if (y < SZ - 1) res += Number(yeetTheNumber(grid, x + 1, y + 1));
               if (y > 0) res += Number(yeetTheNumber(grid, x + 1, y - 1));
               res += Number(yeetTheNumber(grid, x + 1, y));
            }
            if (y > 0) res += Number(yeetTheNumber(grid, x, y - 1));
            if (y < SZ - 1) res += Number(yeetTheNumber(grid, x, y + 1));
         }
      }
   }

   return res.toString();
}

export function part2(input: string, _isTest: boolean): string {
   const grid = input.split('\n').map((s) => s.split(''));
   const SZ = grid.length;

   let res = 0;

   for (let y = 0; y < SZ; y++) {
      for (let x = 0; x < SZ; x++) {
         if (grid[y][x] === '*') {
            let ary = [];
            if (x > 0) {
               if (y < SZ - 1) ary.push(Number(yeetTheNumber(grid, x - 1, y + 1)));
               if (y > 0) ary.push(Number(yeetTheNumber(grid, x - 1, y - 1)));
               ary.push(Number(yeetTheNumber(grid, x - 1, y)));
            }
            if (x < SZ - 1) {
               if (y < SZ - 1) ary.push(Number(yeetTheNumber(grid, x + 1, y + 1)));
               if (y > 0) ary.push(Number(yeetTheNumber(grid, x + 1, y - 1)));
               ary.push(Number(yeetTheNumber(grid, x + 1, y)));
            }
            if (y > 0) ary.push(Number(yeetTheNumber(grid, x, y - 1)));
            if (y < SZ - 1) ary.push(Number(yeetTheNumber(grid, x, y + 1)));
            ary = ary.filter((e) => e);
            if (ary.length === 2) res += ary.reduce((pv, v) => pv * v, 1);
         }
      }
   }

   return res.toString();
}

if (import.meta.main) {
   run(import.meta.url, part1, part2, HAS_ALTERNATE);
}

File: 2023/03/lua/main.lua
#!/usr/local/bin/lua

local run = require('src.lua.run')
local array = require('src.lua.utils.array')
local misc = require('src.lua.utils.misc')

local options = {
   has_alternate = false,
   has_io = false,
}

local function is_symbol(c)
   return (c == '*' or c == '$' or c == '=' or c == '#' or c == '%' or
      c == '/' or c == '&' or c == '+' or c == '-' or c == '@')
end

local function is_num(c)
   return c >= '0' and c <= '9'
end

local function yeet_the_number(grid, x, y)
   local res = ''
   if is_num(grid[y][x]) then
      res = grid[y][x]
      grid[y][x] = '.'
      if x > 1 then
         res = yeet_the_number(grid, x - 1, y) .. res
      end
      if x < #grid[y] then
         res = res .. yeet_the_number(grid, x + 1, y)
      end
   end
   return res
end

local function try_parse_int(s)
   if s == '' then
      return 0
   end
   return tonumber(s)
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local grid = misc.to_grid(input)
   local res = 0

   for y = 1, #grid do
      for x = 1, #grid[y] do
         if is_symbol(grid[y][x]) then
            if x > 1 then
               if y < #grid then
                  res = res + try_parse_int(yeet_the_number(grid, x - 1, y + 1))
               end
               if y > 1 then
                  res = res + try_parse_int(yeet_the_number(grid, x - 1, y - 1))
               end
               res = res + try_parse_int(yeet_the_number(grid, x - 1, y))
            end
            if x < #grid[y] then
               if y < #grid then
                  res = res + try_parse_int(yeet_the_number(grid, x + 1, y + 1))
               end
               if y > 1 then
                  res = res + try_parse_int(yeet_the_number(grid, x + 1, y - 1))
               end
               res = res + try_parse_int(yeet_the_number(grid, x + 1, y))
            end
            if y < #grid then
               res = res + try_parse_int(yeet_the_number(grid, x, y + 1))
            end
            if y > 1 then
               res = res + try_parse_int(yeet_the_number(grid, x, y - 1))
            end
         end
      end
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local grid = misc.to_grid(input)
   local res = 0

   for y = 1, #grid do
      for x = 1, #grid[y] do
         if grid[y][x] == '*' then
            local ary = { 0, 0, 0, 0, 0, 0, 0, 0 }
            if x > 1 then
               if y < #grid then
                  ary[1] = try_parse_int(yeet_the_number(grid, x - 1, y + 1))
               end
               if y > 1 then
                  ary[2] = try_parse_int(yeet_the_number(grid, x - 1, y - 1))
               end
               ary[3] = try_parse_int(yeet_the_number(grid, x - 1, y))
            end
            if x < #grid[y] then
               if y < #grid then
                  ary[4] = try_parse_int(yeet_the_number(grid, x + 1, y + 1))
               end
               if y > 1 then
                  ary[5] = try_parse_int(yeet_the_number(grid, x + 1, y - 1))
               end
               ary[6] = try_parse_int(yeet_the_number(grid, x + 1, y))
            end
            if y < #grid then
               ary[7] = try_parse_int(yeet_the_number(grid, x, y + 1))
            end
            if y > 1 then
               ary[8] = try_parse_int(yeet_the_number(grid, x, y - 1))
            end
            ary = array.filter(ary, function(val) return val ~= 0 end)
            if #ary == 2 then
               res = res + ary[1] * ary[2]
            end
         end
      end
   end

   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()
